crates/modelardb_compression/src/compression.rs

```diff
@@ -35,6 +37,146 @@ use crate::types::{CompressedSegmentBatchBuilder, CompressedSegmentBuilder, Mode
+/// resulting segments.
+pub fn try_compress_multivariate_record_batch(
+    time_series_table_metadata: &TimeSeriesTableMetadata,
+    uncompressed_time_series: &RecordBatch,
+) -> Result<Vec<RecordBatch>> {
+    // Sort by all tags and then time to simplify splitting the data into time series.
+    let sorted_uncompressed_data =
+        sort_record_batch_by_tags_and_time(time_series_table_metadata, uncompressed_time_series)?;
+
+    // Split the sorted uncompressed data into time series and compress them separately.
+    let mut compressed_data = vec![];
+
+    let tag_column_arrays: Vec<&StringArray> = time_series_table_metadata
+        .tag_column_indices
+        .iter()
+        .map(|index| modelardb_types::array!(sorted_uncompressed_data, *index, StringArray))
+        .collect();
+
+    let mut tag_values = Vec::with_capacity(tag_column_arrays.len());
+    for tag_column_array in &tag_column_arrays {
+        tag_values.push(tag_column_array.value(0).to_owned());
+    }
+
+    // The index of the first data point of each time series must be stored so slices
+    // containing only data points for each time series can be extracted and compressed.
+    let mut row_index_start = 0;
+    for row_index in 0..sorted_uncompressed_data.num_rows() {
+        // If any of the tags differ, the data point is from a new time series.
+        let mut is_new_time_series = false;
+        for tag_column_index in 0..tag_column_arrays.len() {
+            is_new_time_series |= tag_values[tag_column_index]
+                != tag_column_arrays[tag_column_index].value(row_index);
+        }
+
+        if is_new_time_series {
+            let time_series_length = row_index - row_index_start;
+            let uncompressed_time_series =
+                sorted_uncompressed_data.slice(row_index_start, time_series_length);
+
+            try_compress_univariate_record_batch(
+                time_series_table_metadata,
+                &uncompressed_time_series,
+                &tag_values,
+                &mut compressed_data,
+            )?;
+
+            for (tag_column_index, tag_column_array) in tag_column_arrays.iter().enumerate() {
+                tag_values[tag_column_index] = tag_column_array.value(row_index).to_owned();
+            }
+
+            row_index_start = row_index;
+        }
+    }
+
+    let time_series_length = sorted_uncompressed_data.num_rows() - row_index_start;
+    let uncompressed_time_series =
+        sorted_uncompressed_data.slice(row_index_start, time_series_length);
+
+    try_compress_univariate_record_batch(
+        time_series_table_metadata,
+        &uncompressed_time_series,
+        &tag_values,
+        &mut compressed_data,
+    )?;
+
+    Ok(compressed_data)
+}
+
+/// Sort the `uncompressed_data` from the time series table with `time_series_table_metadata`
+/// according to its tags and then timestamps.
+fn sort_record_batch_by_tags_and_time(
```

Comment:
if the other function names are changed, maybe also change this to something like `sort_time_series_by_tags_and_time`.